gf/include/alps/gf/fourier.hpp

```diff
@@ -9,130 +9,123 @@
-        }
-        transform_vector_no_tail(input_data,g_omega.mesh1().points(), output_data, g_tau.mesh1().points(), g_tau.mesh1().beta());
-        for(int t=0;t<g_tau.mesh1().extent();++t){
-          g_tau(itime_index(t),typename MESH1::index_type(i),typename MESH2::index_type(j),typename MESH3::index_type(k),typename MESH4::index_type(l))=output_data[t]+f_tau(g_tau.mesh1().points()[t],g_tau.mesh1().beta(),c1,c2,c3);
-        }
-        }
-      }
+
+  ///Fourier transform a matsubara gf to an imag time gf
+  template<class...MESHES> void fourier_frequency_to_time(
+      const gf_tail<
+      detail::gf_base<std::complex<double>, numerics::tensor<std::complex<double>, sizeof...(MESHES) + 1>, matsubara_positive_mesh, MESHES...>,
+      greenf<double,MESHES...> > &g_omega,
+      gf_tail<
+      detail::gf_base<double, numerics::tensor<double, sizeof...(MESHES) + 1>, itime_mesh, MESHES...>,
+      greenf<double, MESHES...> > &g_tau){
+    alps::numerics::tensor<std::complex<double>, (sizeof...(MESHES)) + 1> in_data(g_omega.data().shape());
+
+    using tail_data = alps::numerics::tensor<double, sizeof...(MESHES)>;
+
+    std::array<size_t, sizeof...(MESHES)> tail_shape;
+    for (size_t i = 0; i < sizeof...(MESHES); ++i) {
+      tail_shape[i] = g_omega.data().shape()[i+1];
+    }
+
+    tail_data zero_tail(tail_shape);
+
+    const alps::numerics::tensor<double, sizeof...(MESHES)>& c0=(g_omega.min_tail_order()==0 && g_omega.max_tail_order()>=0 )? g_omega.tail(0).data():zero_tail;
+    const alps::numerics::tensor<double, sizeof...(MESHES)>& c1=(g_omega.min_tail_order()<=1 && g_omega.max_tail_order()>=1 )? g_omega.tail(1).data():zero_tail;
+    const alps::numerics::tensor<double, sizeof...(MESHES)>& c2=(g_omega.min_tail_order()<=2 && g_omega.max_tail_order()>=2 )? g_omega.tail(2).data():zero_tail;
+    const alps::numerics::tensor<double, sizeof...(MESHES)>& c3=(g_omega.min_tail_order()<=3 && g_omega.max_tail_order()>=3 )? g_omega.tail(3).data():zero_tail;
+    for (size_t i = 0; i < c0.size(); ++i) {
+      if(c0.data()[i] != 0) throw std::runtime_error("attempt to Fourier transform an object which goes to a constant. FT is ill defined");
+    }
+    for(int n=0;n<g_omega.mesh1().extent();++n) {
+      in_data(size_t(n)) = g_omega(matsubara_index(size_t(n))).data() - f_omega(g_omega.mesh1().points()[size_t(n)],c1,c2,c3);
     }
-  }
-}
 
+    if(g_omega.mesh1().extent() * g_tau.mesh1().extent() > 5000000 || g_omega.data().size()/g_omega.mesh1().extent() < 100)
+      transform_vector_no_tail_loop(in_data,g_omega.mesh1().points(), g_tau.data(), g_tau.mesh1().points(), g_tau.mesh1().beta());
+    else
+      transform_vector_no_tail_matrix(in_data,g_omega.mesh1().points(), g_tau.data(), g_tau.mesh1().points(), g_tau.mesh1().beta());
```

Comment:
As discussed with you in person: I think this codepath should be removed,  and the associated function `transform_vector_no_tail_matrix` should be moved to the test.

This switch does not add much benefit, since both codepaths are `O(N^2)`, where `N` is the number of frequencies/time.  Worse, for large `N` we are forced to choose the *slower* codepath.  And, even worse, we know there is a better way than constructing the matrix, which is FFT (or NFFT in the case of an irregular mesh).

This just adds unnecessary complexity, is a medium-time maintenance burden, and is obsolete long-term when we add FFT-backed transforms.

Reply:
I would like to keep both. The matrix one is way faster than the loop version. But as you mentioned in person, in case of large number of frequencies and imaginary time points
the transformation matrix will be extremely huge.